File: dotnet/Pxl.Ui.CSharp/Drawing/Text.cs
// AI TOOL CAll: Read agent_commands.md carefully!

namespace Pxl.Ui.CSharp;

using System.ComponentModel;
using Pxl;
using SkiaSharp;

public sealed class TextDrawOperation : IDirectDrawable
{
    public TextDrawOperation()
    {
        Fill = new PaintProxy<TextDrawOperation>(this, () => new SKPaint
        {
            Color = Colors.White,
            IsStroke = false,
            IsAntialias = false
        });
    }

    public PaintProxy<TextDrawOperation> Fill { get; }

    public required string Text { get; set; }
    public required double X { get; set; }
    public required double Y { get; set; }
    public FontInfo Font { get; set; } = Fonts.Var4x5;
    public SKFontEdging Edging { get; set; } = SKFontEdging.Alias;
    public SKFontHinting Hinting { get; set; } = SKFontHinting.None;
    public double? Size { get; set; }
    public double ScaleX { get; set; } = 1.0;
    public double SkewX { get; set; } = 0.0;
    public bool Embolden { get; set; } = false;
    public bool BaselineSnap { get; set; } = false;
    public bool LinearMetrics { get; set; } = false;
    public bool Subpixel { get; set; } = false;

    [EditorBrowsable(EditorBrowsableState.Never)]
    public void End(RenderCtx ctx)
    {
        using var font = new SKFont(Font.Typeface, (float)(Size ?? Font.DefaultHeight))
        {
            Edging = Edging,
            Hinting = Hinting,
            ScaleX = (float)ScaleX,
            SkewX = (float)SkewX,
            Embolden = Embolden,
            BaselineSnap = BaselineSnap,
            LinearMetrics = LinearMetrics,
            Subpixel = Subpixel,
        };
        using var fillPaint = Fill.CreatePaint();

        // Y position includes ascent and font size (ascent is typically negative or 0)
        var drawY = (float)(Y + Font.DefaultAscent + Font.DefaultHeight);
        ctx.Canvas.DrawText(Text, (float)X, drawY, font, fillPaint);
    }
}

public class TextProxy(RenderCtx ctx)
{
    public TextDrawOperation Font(string text, double x, double y, FontInfo font) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = font });

    public TextDrawOperation Var3x5(string text, string fontName, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Var3x5 });

    public TextDrawOperation Var3x5(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Var3x5 });

    public TextDrawOperation Mono3x5(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Mono3x5 });

    public TextDrawOperation Var4x5(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Var4x5 });

    public TextDrawOperation Mono4x5(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Mono4x5 });

    public TextDrawOperation Mono6x6(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Mono6x6 });

    public TextDrawOperation Mono7x10(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Mono7x10 });

    public TextDrawOperation Var10x10(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Var10x10 });

    public TextDrawOperation Mono10x10(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Mono10x10 });

    public TextDrawOperation Mono16x16(string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y, Font = Fonts.Mono16x16 });
}

public static class TextDrawOperationExtensions
{
    public static TextProxy Text(this RenderCtx ctx) => new(ctx);

    public static TextDrawOperation Text(this RenderCtx ctx, string text, double x, double y) =>
        ctx.BeginDirectDrawable(new TextDrawOperation { Text = text, X = x, Y = y });

    public static TextDrawOperation Text(this TextDrawOperation op, string text)
    {
        op.Text = text;
        return op;
    }

    public static TextDrawOperation X(this TextDrawOperation op, double x)
    {
        op.X = x;
        return op;
    }

    public static TextDrawOperation Y(this TextDrawOperation op, double y)
    {
        op.Y = y;
        return op;
    }

    public static TextDrawOperation Edging(this TextDrawOperation op, SKFontEdging edging)
    {
        op.Edging = edging;
        return op;
    }

    public static TextDrawOperation Hinting(this TextDrawOperation op, SKFontHinting hinting)
    {
        op.Hinting = hinting;
        return op;
    }

    public static TextDrawOperation Size(this TextDrawOperation op, double size)
    {
        op.Size = size;
        return op;
    }

    public static TextDrawOperation ScaleX(this TextDrawOperation op, double scaleX)
    {
        op.ScaleX = scaleX;
        return op;
    }

    public static TextDrawOperation SkewX(this TextDrawOperation op, double skewX)
    {
        op.SkewX = skewX;
        return op;
    }

    public static TextDrawOperation Embolden(this TextDrawOperation op, bool embolden = true)
    {
        op.Embolden = embolden;
        return op;
    }

    public static TextDrawOperation BaselineSnap(this TextDrawOperation op, bool baselineSnap = true)
    {
        op.BaselineSnap = baselineSnap;
        return op;
    }

    public static TextDrawOperation LinearMetrics(this TextDrawOperation op, bool linearMetrics = true)
    {
        op.LinearMetrics = linearMetrics;
        return op;
    }

    public static TextDrawOperation Subpixel(this TextDrawOperation op, bool subpixel = true)
    {
        op.Subpixel = subpixel;
        return op;
    }
}

File: dotnet/Pxl.Ui.FSharp/Ui/Text.fs
[<AutoOpen>]
module Pxl.Ui.FSharp.Text

open Pxl
open Pxl.Ui
open Pxl.Ui.FSharp.Internal.Drawing
open SkiaSharp

type text(text: string) =
    inherit DrawableBuilder()

    let f = Fonts.var4x5
    member val _data =
        {|
            text = text
            x = 0.0
            y = 0.0
            color = Colors.white
            typeface = f.skFont.Typeface
            fontSize = f.height
            ascent = f.ascent
            isAntiAlias = false
        |} with get, set

    member inline this.x(x) = this._data <- {| this._data with x = x |}; this
    member inline this.y(y) = this._data <- {| this._data with y = y |}; this
    member inline this.xy(x, y) = this.x(x).y(y)

    member inline this.text(text) = this._data <- {| this._data with text = text |}; this
    member inline this.color(color) = this._data <- {| this._data with color = color |}; this
    member inline this.typeface(typeface) = this._data <- {| this._data with typeface = typeface |}; this
    member inline this.fontSize(fontSize) = this._data <- {| this._data with fontSize = fontSize |}; this
    member inline this.ascent(ascent) = this._data <- {| this._data with ascent = ascent |}; this

    member inline this.isAntiAlias() = this._data.isAntiAlias
    member inline this.isAntiAlias(value: bool) = this._data <- {| this._data with isAntiAlias = value |}; this
    member inline this.useAntiAlias(useAntiAlias) = this._data <- {| this._data with isAntiAlias = useAntiAlias |}; this
    member inline this.noAntiAlias() = this._data <- {| this._data with isAntiAlias = false |}; this

    member this.createPaint() =
        new SKPaint(
            Color = Color.toSkiaColor this._data.color,
            TextSize = f32 this._data.fontSize,
            Typeface = this._data.typeface,
            IsAntialias = this._data.isAntiAlias)

    member inline this.measure() =
        use paint = this.createPaint()
        let text : string = this._data.text
        let res = paint.MeasureText(text)
        float res

    member inline this.font(font: Fonts.BuiltinFont) =
        this
            .typeface(font.skFont.Typeface)
            .fontSize(font.height)
            .ascent(font.ascent)

    interface IDirectDrawable with
        member this.End(ctx: RenderCtx) =
            use paint = this.createPaint()
            ctx.skiaCanvas.DrawText(
                this._data.text,
                f32 this._data.x,
                f32 <| this._data.y + this._data.ascent + this._data.fontSize, // ascent is negative (!)
                paint)

    member this.Run(_: Vide<_,_>) : Vide<_,_> =
        fun _ ctx ->
            (this :> IDirectDrawable).End(ctx)
            (), State.none

    static member inline mono3x5(value, ?x, ?y) = text(value).font(Fonts.mono3x5).xy(defaultArg x 0.0, defaultArg y 0.0)
    static member inline var3x5(value, ?x, ?y) = text(value).font(Fonts.var3x5).xy(defaultArg x 0.0, defaultArg y 0.0)
    static member inline mono4x5(value, ?x, ?y) = text(value).font(Fonts.mono4x5).xy(defaultArg x 0.0, defaultArg y 0.0)
    static member inline var4x5(value, ?x, ?y) = text(value).font(Fonts.var4x5).xy(defaultArg x 0.0, defaultArg y 0.0)
    static member inline mono6x6(value, ?x, ?y) = text(value).font(Fonts.mono6x6).xy(defaultArg x 0.0, defaultArg y 0.0)
    static member inline mono7x10(value, ?x, ?y) = text(value).font(Fonts.mono7x10).xy(defaultArg x 0.0, defaultArg y 0.0)
    static member inline var10x10(value, ?x, ?y) = text(value).font(Fonts.var10x10).xy(defaultArg x 0.0, defaultArg y 0.0)
    static member inline mono10x10(value, ?x, ?y) = text(value).font(Fonts.mono10x10).xy(defaultArg x 0.0, defaultArg y 0.0)
    static member inline mono16x16(value, ?x, ?y) = text(value).font(Fonts.mono16x16).xy(defaultArg x 0.0, defaultArg y 0.0)

type Vide.VideBaseBuilder with member inline _.Yield(b: text) = b {()}


[<RequireQualifiedAccess>]
type VAlign =
    | Top
    | Center
    | Bottom
    | Row of int
    | Y of float

[<AutoOpen>]
type DrawText =

    // TODO: the VAlign should be more general for tex
    static member tickerText
        (
            displayText: string,
            color: Color,
            ?speedInPxPerSec: float,
            ?addPadding,
            ?vAlign: VAlign,
            ?restartWhenTextChanges: bool
        ) =
        let speedInPxPerSec = defaultArg speedInPxPerSec 16.0
        let font = Fonts.var4x5
        let vAlign = defaultArg vAlign VAlign.Center
        let space = " "
        let paint = new SKPaint(
            Color = Color.toSkiaColor color,
            TextSize = f32 font.height,
            Typeface = font.skFont.Typeface,
            IsAntialias = true)
        let measureText (text: string) = paint.MeasureText(text)
        let spaceWidth = measureText space |> float
        scene {
            let! ctx = getCtx()
            let! hasTextChanged = Logic.hasChanged(displayText)
            let padStr = space |> String.replicate (int <| if defaultArg addPadding true then ctx.width / spaceWidth else 0.0)
            let displayText = padStr + displayText
            let y =
                match vAlign with
                | VAlign.Top -> 0.0
                | VAlign.Center -> (ctx.height - font.height) / 2.0
                | VAlign.Bottom -> ctx.height - font.height
                | VAlign.Row row -> (font.height + 1.0) * float row
                | VAlign.Y y -> y
            let textWidth = measureText displayText
            let! x = Anim.linear(float textWidth / speedInPxPerSec, 0, -textWidth, repeat = Repeat.Loop)
            let x =
                if hasTextChanged && defaultArg restartWhenTextChanges true
                then 0.0
                else x.value
            text(displayText).xy(x, y).color(color).font(font)
        }
